### PR TITLE
Optimize system & Overhaul code ;)

### PR DESCRIPTION
**Pull request by ghost**
- Demon invasions can increase up to 6 towards the later stages (aka all running), so setup frame.bar to be a table to contain all them, even ones not setup yet (just leave them unused)
- Instead of stopping the bar (and deleting it in the process), simply just reuse the bar by updating its duration (and re-starting to resetup all the data)
- Optimize bar loading into a function, and loop (save coding it manually every time)
- Use a select function for scenarioType, save many underscores
- For LibCandyBar_Stop, it'll now only be called when the timer expires, therefor a new invasion has begun
- timeLeftMinutes now no longer reports more than the max time (Blizzard fixed)
- Don't bother checking for justLoggedIn, as if the event is new, alert them either way
- Optimize the f.db and f.header values (don't create a local, just init them directly)
- No longer have a paused variable for startTimer, just detect if achievement ID is nil (also use nil rather than 0, quicker checking)

### CODE REVIEW

**Comment by funkydude** [comment]
Thanks for taking the time on this. Unfortunately a lot of the changes are actually performance downgrades, I'll go through everything individually and comment on it, although the extra bar support will save me the time implementing, thanks for that!

Whilst I comment on things can you fill me in on this? 

> timeLeftMinutes now no longer reports more than the max time (Blizzard fixed)

How do you know this for sure, the issue only affected certain servers, I could never reproduce it myself.

File: LegionInvasionTimer.lua [hunk 1]
```diff
@@ -186,43 +159,30 @@ frame:SetScript("OnEvent", function(f)
...
-			frame.bar3 = nil
-		end
 	end)
 
 	findTimer()
 	f:RegisterEvent("SCENARIO_COMPLETED")
 	f:SetScript("OnEvent", function()
-		local _,_,_,_,_,_,_,_,_,scenarioType = C_Scenario.GetInfo()
-		if scenarioType == 4 then -- LE_SCENARIO_TYPE_LEGION_INVASION = 4
+		if select(10, C_Scenario.GetInfo()) == 4 then -- LE_SCENARIO_TYPE_LEGION_INVASION = 4
```
Comment (funkydude) [inline LegionInvasionTimer.lua]: This is less efficient and is actually a downgrade, all it does is make "cleaner" code, however, I do not code my addons for clean code, I code them for performance.

File: LegionInvasionTimer.lua [hunk 2]
```diff
@@ -186,43 +159,30 @@ frame:SetScript("OnEvent", function(f)
...
-	bg:SetAllPoints(f)
-	bg:SetColorTexture(0, 1, 0, 0.3)
-	f.bg = bg
-	local header = f:CreateFontString(nil, "OVERLAY", "TextStatusBarText")
-	header:SetAllPoints(f)
-	header:SetText(name)
-	f.header = header
-	f.startBar = startBar
-	f.db = legionTimerDB
+	f.bg = f:CreateTexture(nil, "PARENT")
```
Comment (funkydude) [inline LegionInvasionTimer.lua]: This is an unnecessary change that produces a minuscule performance downgrade. This method forces a table lookup every single time a change to a widget is made. If you are accessing a table entry multiple times, it is more efficient to do so via a local variable.

File: LegionInvasionTimer.lua [hunk 3]
```diff
@@ -186,43 +159,30 @@ frame:SetScript("OnEvent", function(f)
...
 
 	if legionTimerDB.lock then
 		f:EnableMouse(false)
 		f.bg:Hide()
 		f.header:Hide()
 	end
 
-	candy.RegisterCallback(name, "LibCandyBar_Stop", function(_, bar, dontScan)
-		if not dontScan and bar == frame.bar1 and bar:Get("LegionInvasionTimer:complete") then
+	candy.RegisterCallback(name, "LibCandyBar_Stop", function(_, bar)
```
Comment (funkydude) [inline LegionInvasionTimer.lua:185]: Your current implementation piles the boss bars (which are now broken by this commit) onto the active zone duration bars by removing code like this, I'd rather not do that. Recycling the zone bars keeps the screen cleaner and keep the most relevant information showing.

Comment (ghost) [inline LegionInvasionTimer.lua:185]: Then surely it's worth more actually having the boss abilities being handled via DBM.

Or, create new bars for the bosses (and a config option to disable them, for those who also have DBM installed)

**Comment by ghost** [comment]
> How do you know this for sure, the issue only affected certain servers, I could never reproduce it myself.

Blizzard hotfixes for 11th August:
"Fixed a bug that caused the time remaining for some invasions to display an incorrect value." (https://worldofwarcraft.com/en-us/news/20195220/patch-hotfixes-august-12)

**Comment by funkydude** [comment]
> Blizzard hotfixes for 11th August:

It's possible the two are related, but the fix was on the 12th when the issue was actively happening: https://github.com/funkydude/LegionInvasionTimer/commit/a5987c3501c772bc93b5649b52612becd463fc0b

I guess we could try and see if it's fixed, maybe the hotfixed was delayed.

File: LegionInvasionTimer.lua (hunk 3 above)
Comment (funkydude) [inline LegionInvasionTimer.lua:185]: Why should users need to install inefficient junk like DBM to have a few warnings for a temporary event? Not to mention this covers more warnings with more accurate timers.

Sorry, I'm just not interested in removing the boss mods :(

**Comment by ghost** [comment]
Well if there's a new report for the issue still occurring, I can easily hop on to the realm and do some debugging to see what's actually happening (Technically, the API data should be returning correctly)

File: LegionInvasionTimer.lua (hunk 3 above)
Comment (ghost) [inline LegionInvasionTimer.lua:185]: Wasn't saying to remove. I was saying to firstly make it a config option (make it loadable like options), and to secondly have it's own set of timers (so that them timers don't mix with LegionInvasionTimer bars for zones). That way it doesn't effect those who DO have DBM installed, and allow them to turn it off.

Comment (funkydude) [inline LegionInvasionTimer.lua:185]: There's already an option to turn it off, I don't see a reason to create a separate anchor and bars for it.

Comment (ghost) [inline LegionInvasionTimer.lua:185]: I'll look into updating the bosses system later :) So PR on hold till that happens

**Comment by funkydude** [comment]
Gonna close this. The main upgrade here is definitely the proper bar support instead of the current static 3 bars. I'll introduce this soon using your work and give you credit, thanks a lot for your time on this!
